### PR TITLE
Improve divisible split detection

### PR DESCRIPTION
**Pull request by csarofeen**
Adds propagation of divisible split information, as well as add divisible splits from view based transformations.

Tangibly NVFuserTest.FusionNonDivisibleSplitVectorize2_CUDA before:
```.cpp
__global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 0> T3) {
  alignas(16) extern __shared__ char array[];
  void* shared_mem = array;
  NVFUSER_DEFINE_MAGIC_ZERO
  Array<float, (8 * 4), 4> T1;
  #pragma unroll
  for(nvfuser_index_t i21 = 0; i21 < 8; ++i21) {
    if (((((i21 + nvfuser_zero) * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + 3)) < T0.size[0])) {
      loadGlobalToLocal<float, 4, false>(&T1[(i21 * 4)],  &T0[(((i21 + nvfuser_zero) * (ceilDiv(T0.size[0], 8))) + (((nvfuser_index_t)threadIdx.x) * 4))]);
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  // Alias Allocation - register
  auto& T2 = T1;
  #pragma unroll
  for(nvfuser_index_t i23 = 0; i23 < 8; ++i23) {
    #pragma unroll
    for(nvfuser_index_t i24 = 0; i24 < 4; ++i24) {
      if ((((i23 * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + (i24 + nvfuser_zero))) < T0.size[0])) {
        T2[((i23 * 4) + i24)]
          = T1[((i23 * 4) + i24)]
          + (float) 1.00000000000000000e+00;
      }
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  T3[0] = 0.00000000000000000e+00;
  #pragma unroll
  for(nvfuser_index_t i25 = 0; i25 < 8; ++i25) {
    #pragma unroll
    for(nvfuser_index_t i26 = 0; i26 < 4; ++i26) {
      blockReduce<true, false, false>(
        T3[0],
        T2[((i25 * 4) + i26)],
        [](float &a, float b) { a = a + b; },
        threadIdx,
        blockDim,
        static_cast<float*>(shared_mem),
        (((i25 * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + (i26 + nvfuser_zero))) < T0.size[0]),
        float(0.00000000000000000e+00));
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
}
```
After:
```.cpp
__global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 0> T3) {
  alignas(16) extern __shared__ char array[];
  void* shared_mem = array;
  NVFUSER_DEFINE_MAGIC_ZERO
  Array<float, (8 * 4), 4> T1;
  #pragma unroll
  for(nvfuser_index_t i21 = 0; i21 < 8; ++i21) {
    T1.set(0);
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  #pragma unroll
  for(nvfuser_index_t i21 = 0; i21 < 8; ++i21) {
    if (((((i21 + nvfuser_zero) * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + 3)) < T0.size[0])) {
      loadGlobalToLocal<float, 4, false>(&T1[(i21 * 4)],  &T0[(((i21 + nvfuser_zero) * (ceilDiv(T0.size[0], 8))) + (((nvfuser_index_t)threadIdx.x) * 4))]);
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  // Alias Allocation - register
  auto& T2 = T1;
  #pragma unroll
  for(nvfuser_index_t i23 = 0; i23 < 8; ++i23) {
    #pragma unroll
    for(nvfuser_index_t i24 = 0; i24 < 4; ++i24) {
      T2[((i23 * 4) + i24)]
        = T1[((i23 * 4) + i24)]
        + (float) 1.00000000000000000e+00;
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
  T3[0] = 0.00000000000000000e+00;
  #pragma unroll
  for(nvfuser_index_t i25 = 0; i25 < 8; ++i25) {
    #pragma unroll
    for(nvfuser_index_t i26 = 0; i26 < 4; ++i26) {
      blockReduce<true, false, false>(
        T3[0],
        T2[((i25 * 4) + i26)],
        [](float &a, float b) { a = a + b; },
        threadIdx,
        blockDim,
        static_cast<float*>(shared_mem),
        (((i25 * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + (i26 + nvfuser_zero))) < T0.size[0]),
        float(0.00000000000000000e+00));
    }
  }
  NVFUSER_UPDATE_MAGIC_ZERO
}
```

Specifically this predicate is successfully removed:
```
      if ((((i23 * (ceilDiv(T0.size[0], 8))) + ((((nvfuser_index_t)threadIdx.x) * 4) + (i24 + nvfuser_zero))) < T0.size[0])) {
        T2[((i23 * 4) + i24)]
          = T1[((i23 * 4) + i24)]
          + (float) 1.00000000000000000e+00;
      }
```

### CODE REVIEW

**Comment by csarofeen** [comment]
@naoyam or @zasdfgbnm this should be ready to review. There is one bug here, which is that rfactor domains that are a result of non-divisible splits (only happens with reduction) might not be identified as such, but indexed as such. Was thinking @samnordmann might be able to help here, since this would be required to fix if we wanted mult-gpu supported rfactor stages.

File: torch/csrc/jit/codegen/cuda/lower_divisible_split.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+// Looks through all transformations assocaited with view, or enforced divisible
+// vectorization splits and gathers all splits that provably don't have a
+// remainder, therefore the extents of the associated IterDomains do not require
+// a ceilDiv expressions.
+TORCH_CUDA_CU_API std::unordered_set<Split*> getAllDivisibleSplits(
+    Fusion* fusion);
+
+// Same as above but will use provided ComputeAtMap instead of building its own.
+std::unordered_set<Split*> getAllDivisibleSplits(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisible_split.h]: nit: any specific reason not to have `TORCH_CUDA_CU_API` for this version.

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: The file name should be `lower_divisible_split.cpp`.

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    auto rfactor_dom = tv->getMaybeRFactorDomain();
+    // Not view if there's no rfactor axis
+    if (!tv->hasRFactor() ||
+        std::any_of(
+            rfactor_dom.begin(),
+            rfactor_dom.end(),
+            // Also not a view transform if there's a reduction dimension.
+            [](IterDomain* id) { return id->isReduction(); })) {
+      continue;
+    }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: Can this be simplified by using `TensorDomain::hasViewLikeRFactor()`?

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+std::unordered_set<Split*> getAllDivisibleSplits(
+    Fusion* fusion,
+    const ComputeAtMap* ca_map) {
+  std::unordered_set<Split*> all_divisible_splits;
+
+  auto all_tvs = ir_utils::allTvs(fusion);
+  // Find all tensor views with a view like rfactor
+  for (auto tv : all_tvs) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: Can you please add a note saying splits used in view are by definition divisible?

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        // Also not a view transform if there's a reduction dimension.
+        [](IterDomain* id) {
+          return isParallelTypeVectorize(id->getParallelType());
+        });
+
+    if (vec_id_it == tv->domain()->domain().end()) {
+      continue;
+    }
+
+    // We could have a cass technically like:
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: type: case

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        StmtSort::getExprs(fusion, {rfactor_dom.begin(), rfactor_dom.end()});
+    auto split_exprs = ir_utils::filterByType<Split>(view_exprs);
+    all_divisible_splits.insert(split_exprs.begin(), split_exprs.end());
+  }
+
+  for (auto tv : all_tvs) {
+    auto vec_id_it = std::find_if(
+        tv->domain()->domain().begin(),
+        tv->domain()->domain().end(),
+        // Also not a view transform if there's a reduction dimension.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: Unrelated comment

File: torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+  std::unordered_set<Expr*> visited(
+      all_divisible_splits.begin(), all_divisible_splits.end());
+
+  // Find splits that match what we already have:
+  for (auto entry : outer_concrete_id_to_expr) {
+    auto concrete_id = entry.first;
+    auto original_view_split = entry.second;
+
+    auto exact_mapped_ids =
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_divisble_split.cpp]: nit: `const auto&`

File: torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp [hunk 8]
```diff
@@ -1782,6 +1783,67 @@ TEST_F(NVFuserTest, FusionViewMapping_CUDA) {
...
+    auto transform_0 = tv->axis(0)->definition()->as<Split>();
+    auto transform_1 = transform_0->in()->definition()->as<Split>();
+    auto transform_2 = transform_1->in()->definition()->as<Merge>();
+    auto transform_3 = transform_2->outer()->definition()->as<Merge>();
+
+    auto transform_5 = transform_3->inner()->definition()->as<Split>();
+    auto transform_6 = transform_5->in()->definition()->as<Split>();
+    auto transform_7 = transform_6->in()->definition();
+
+    TORCH_CHECK(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:1837]: Why not asserting `transform_6` and `transform_7` are also detected as divisible?

**Review by naoyam — COMMENTED** [review]
Added some comments. Nothing blocking

**Comment by naoyam** [comment]
> @naoyam or @zasdfgbnm this should be ready to review. There is one bug here, which is that rfactor domains that are a result of non-divisible splits (only happens with reduction) might not be identified as such, but indexed as such. Was thinking @samnordmann might be able to help here, since this would be required to fix if we wanted mult-gpu supported rfactor stages.

Is the bug due to this PR? Or is it already in the code?

File: torch/csrc/jit/codegen/cuda/lower_divisible_split.h (hunk 1 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_divisible_split.h]: Not really, I typically only add `TORCH_CUDA_CU_API` when I build a test on an interface. Can add it for symmetry.

**Comment by csarofeen** [comment]
The bug already exists before this PR.

**Review by naoyam — APPROVED** [review]
LGTM.

**Comment by naoyam** [comment]
> The bug already exists before this PR.

Can you please file an issue?

**Comment by csarofeen** [comment]
Yeah, I'm actually going to ask @samnordmann to file an issue on it and work on it.

**Pull request merged**